### PR TITLE
Af ar sharelist

### LINKED ISSUE #5: 5. As a user, I want to join an existing shopping list so I can share a shopping list with another person
## Summary
We want to allow users to join existing shopping lists so they can manage them with friends or family. This would take the following steps:
1. User enters an existing token into a form
2. The token is saved to `localStorage`
3. The user is taken to the `List` view and shown the existing list.

## Acceptance criteria
If a user **doesn’t** already have a token:
- [x] The `Home` view shows a form that allows the user to enter a token to join an _existing_ list, in addition to the button that allows them to create a _new_ list.
- [x] The input that accepts the list token has a semantic `label` element associated with its
- [x] The user can submit this form with both the mouse and the `Enter` key
- [x] If the list exists,
	- the token is saved in `localStorage`
	- the user is redirected to the `List` view and shown the items on that list
- [x] If the list _does not_ exist, the user is shown an error message that explains the problem

If a user **does** already have a token:
- [x] They are automatically redirected to the `List` view.

## Notes
- Be sure to use the `setListToken` function to save the token that the user enters!
- You will know your `<label>` is correctly implemented if you click on the label and keyboard focus moves to the related input. Refer to [the MDN docs on `<label>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/label) for more info.
- Users will be able to submit the form with the `Enter` key if you listen for the right JavaScript event on the right element! If you're adding an `onKeyDown` listener anywhere, you might be overcomplicating things.
- You can assume that users will only be sharing a token if the list has actual items on it. So, let's say I create a New Shopping List, but never add any items to it, and haven't saved that particular token to Firestore, we can assume that's a list that doesn't exist.
- If you would like, you can save the token to Firestore immediately when creating a list, and then a user would be able to share an empty list (see issue 3).

### PR DESCRIPTION
**Pull request by a-fink**
## Description
Added functionality to allow users to share lists, as well as a check that will alert the user if they are trying to join a list that doesn't exist. Allowed users to view their list token for sharing from the List component. 

Due to the way that Firebase handles collections, we decided to keep the placeholder value set up in issue 3 so that users can share empty lists without errors. 

Corrected a bug in filtering out hidden items -> items that were added before the hidden property was set up were being unintentionally filtered out when hidden was undefined.

## Related Issue

Closes #5 

## Acceptance Criteria

If a user **doesn’t** already have a token:
- [x] The `Home` view shows a form that allows the user to enter a token to join an _existing_ list, in addition to the button that allows them to create a _new_ list.
- [x] The input that accepts the list token has a semantic `label` element associated with its
- [x] The user can submit this form with both the mouse and the `Enter` key
- [x] If the list exists,
	- the token is saved in `localStorage`
	- the user is redirected to the `List` view and shown the items on that list
- [x] If the list _does not_ exist, the user is shown an error message that explains the problem

If a user **does** already have a token:
- [x] They are automatically redirected to the `List` view.

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
| ✓   | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
| ✓   | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

### Before

home page before
list before

### After

home page after
list after

## Testing Steps / QA Criteria
- pull down the branch
-  run `npm start` to launch the frontned and navigate to localhost:3000/
- in DevTools under storage, if a 'tcl-shopping-list-token' entry exists in localStorage, delete it and refresh the page
- navigate to the homepage and the new join a list input field and button should appear
- enter the token for an existing list and you should be redirected to the list page and see the list items and the list token displayed
- enter the token for a non-existent list and you should receive a pop-up error message

### CODE REVIEW

**Review by Amanda2900 — APPROVED** [review]
Tested both inputting an incorrect token to get the popup error message and entering a correct one to access that list. Everything works as expected. Nice work to both of you!!

**Review by ameliasheppy — APPROVED** [review]
Great job on it, @a-fink and @alejandronanez !!

File: src/api/firebase.js [hunk 1]
```diff
@@ -1,7 +1,24 @@
...
+} from 'firebase/firestore';
 import { db } from './config';
 import { getFutureDate } from '../utils';
 
+/**
+ * Return whether a collection has documents in it (list token exists) or is empty (list token does not exist)
+ * @param {string} listId the listId that a user submitted to join a list
+ * @returns {boolean}
+ */
+export async function doesCollectionExist(listId) {
```
Comment (luisaugusto) [inline src/api/firebase.js:16]: Nice work building this out!

File: src/api/firebase.js [hunk 2]
```diff
@@ -42,10 +60,10 @@ export function getItemData(snapshot) {
 
 		return data;
 	});
-
+	// console.log(arrayFromFirestore,'arrayFromFirestore')
```
Comment (luisaugusto) [inline src/api/firebase.js]: I think this commented out console log can be removed
```suggestion
```

File: src/api/firebase.js [hunk 3]
```diff
@@ -11,6 +28,7 @@ import { getFutureDate } from '../utils';
  * @see https://firebase.google.com/docs/firestore/query-data/listen
  */
 export function streamListItems(listId, handleSuccess) {
+	// TODO: check if the token is a valid token in firebase
```
Comment (luisaugusto) [inline src/api/firebase.js]: Is this todo something that you'd like to complete in this PR or is that what `doesCollectionExist` is solving above?

File: src/views/Home.jsx [hunk 4]
```diff
@@ -15,10 +18,48 @@ export function Home({ setListToken }) {
...
 	};
 
+	const tokenHandler = (e) => {
+		setTokenInput(e.target.value);
+	};
+
+	const handleTokenSubmit = async (e) => {
+		e.preventDefault();
+		const exists = await doesCollectionExist(tokenInput);
+		console.log(exists);
```
Comment (luisaugusto) [inline src/views/Home.jsx]:
```suggestion
```

**Review by luisaugusto — APPROVED** [review]
Nice work! I like the way you are checking if a list exists before adding to the local storage.

File: src/api/firebase.js (hunk 1 above)
Comment (a-fink) [reply src/api/firebase.js:16]: Thanks! Took us a while to figure out how to get Firebase to let us know if a list exists or not! @Kmachappy was a champ on this one, and kept us both motivated to keep digging until we found a clean answer.

File: src/api/firebase.js (hunk 2 above)
Comment (a-fink) [reply src/api/firebase.js]: Cleaned up in my recent commit, thanks for the catch.

File: src/api/firebase.js (hunk 3 above)
Comment (a-fink) [reply src/api/firebase.js]: It is covered in the doesCollectionExist, comment just stuck around. Cleaned up in my recent commit. Thanks!

File: src/views/Home.jsx (hunk 4 above)
Comment (a-fink) [reply src/views/Home.jsx]: Cleaned up in my recent commit. Thanks!

File: src/api/firebase.js (hunk 1 above)
Comment (mentalcaries) [reply src/api/firebase.js:16]: Brilliant! Nice work, team!

Comment (SinfulHymn) [reply src/api/firebase.js:16]: @a-fink All credits to you for finding the right function, we both had a clear idea what direction to go.

**Pull request merged**